File: bls-server/src/shared/utils/pagination.ts
export interface PageQuery {
  pageNum?: number | string;
  pageSize?: number | string;
}

export interface PageParams {
  pageNum: number;
  pageSize: number;
  offset: number;
}

export function getPageParams(query: PageQuery): PageParams {
  const pageNum = Math.max(Number(query.pageNum ?? 1), 1);
  const pageSize = Math.min(Math.max(Number(query.pageSize ?? 10), 1), 200);
  return { pageNum, pageSize, offset: (pageNum - 1) * pageSize };
}

File: bls-rust-server/src/utils/pagination.rs
use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
pub struct PageParams {
    #[serde(default = "default_page_num")]
    pub page_num: u64,
    #[serde(default = "default_page_size")]
    pub page_size: u64,
    pub keyword: Option<String>,
}

fn default_page_num() -> u64 {
    1
}
fn default_page_size() -> u64 {
    10
}

impl PageParams {
    pub fn offset(&self) -> u64 {
        (self.page_num.max(1) - 1) * self.limit()
    }

    pub fn limit(&self) -> u64 {
        self.page_size.clamp(1, 100)
    }
}
